Modify for gulp --tasks

sttk opened the pull request:
I tried to achieve @tylersticka's  second image in gulpjs/gulp#1294 and finished.

  * Don't hide series/parallel tasks.
  * Can change task tree depth by '--depth' option.
  * Can set a task description with .description property, as follows:   
    ```
       gulp.task('task-name', fn).description = 'task-description';
    ```  
    or  
    ```
       gulp.task('task-name', fn).description = { '': 'task-description', '--option1': 'task-option-1', ...  };
    ```

sttk commented:
Sorry for some errors but I fixed them all.

phated commented:
Many of the changes in here aren't required fir the CLI changes, including the description changes and the `task` function changes.  The transformer code is interesting but very complex and possibly not needed.

phated commented:
To clarify, the `undertaker.task()` method is a setter and a getter for tasks, it should not return a task when it is being set.

How users should set description:
```js
function clean(cb){ cb() }
clean.description = 'remove some folders';
gulp.task(clean);
```

sttk commented:
> The transformer code is interesting but very complex and possibly not needed.

I got it. I'll try to remove the transformer from undertaker#tree and not use it in gulp-cli#logTasks.

sttk commented:
> How users should set description:

Isn't there a case like the following?

```
function clean(cb) { cb() }
clean.description = 'remove some folders';
gulp.task(clean);
gulp.task('another task name', clean).description = 'another description';
```

Or don't you want to write like the following?

```
gulp.task('clean', function(cb) { cb() })
 .description = 'remove some folders';
```

sttk commented:
How about the depth flag and the sorting of undertaker#tree. Should the sorting be optional?

phated commented:
I'd prefer not to allow chaining description because that requires the setter to return the function. If you are able to reduce the transformer code, I'm good with leaving it in here. Not sure about the sorting but you can use lodash sortBy for it

phated commented:
I'd put depth in the options, I think.

sttk commented:
I've finished modifying the program!

> I'd prefer not to allow chaining description because that requires the setter to return the function. 

OK. I removed the chaining description. 
However it's not difficult to implement it again, so I may do so if you or anyone want it later.

> Not sure about the sorting but you can use lodash sortBy for it 

Yes, I used it in my code. It is easy to write as needed, so I removed the sorting from undertaker#tree not to change the original behavior preferably.

phated commented on lib/tree.js:
```diff
@@ -7,18 +7,36 @@ var metadata = require('./helpers/metadata');
+    });
+  } else {
+    var maxDepth = (opts.depth == null || opts.depth > 0) ? opts.depth : 1;
+    nodes = _.map(this._registry.tasks(), function(task) {
```
you can use the `tasks` variable from above

sttk replied:
Thank you for pointing out. I'll modify it.

phated commented on lib/tree.js:
```diff
@@ -7,18 +7,36 @@ var metadata = require('./helpers/metadata');
-    return meta.tree.label;
-  });
+  var nodes;
+  if (!opts.deep) {
```
can you normalize `opts.deep` to `depth: Infinity` and avoid the branching logic?  Deep can be there for backwards compat or we can just remove it and release as a breaking change (this module isn't 1.0 yet)

sttk replied:
Is it ok that `opts.deep` should be removed and show all depth when `opts.depth == null || opts.depth == Infinity` ?

phated commented on lib/tree.js:
```diff
@@ -7,18 +7,36 @@ var metadata = require('./helpers/metadata');
+        return node;
+      }
+      if (meta.orig) {
+        if (meta.orig.description) {
```
can these be retrieved inside gulp-cli?  were you just trying to reduce the times we iterate over the tasks list?  I'd prefer `description` and `flags` not to mean anything to undertaker since they are a very gulp-cli specific thing

sttk replied:
OK. I'll try to leave only the codes for depth and move the codes to operating `.description` and `.flags`.

sttk replied:
Or, is it better that I'll move even the codes for depth to gulp-cli to reduce the times of task listing? If so, the change in undertaker will be only removing `opts.deep`.

phated commented on lib/tree.js:
```diff
@@ -7,18 +7,36 @@ var metadata = require('./helpers/metadata');
+        if (meta.orig.description) {
+          node.description = meta.orig.description;
+        }
+        if (meta.orig.flag) {
```
name this `flags` instead of the singular `flag`

sttk replied:
OK. I'll modify so.

phated commented on test/task.js:
```diff
@@ -24,21 +24,28 @@ describe('task', function() {
 
   it('should register a named function', function(done) {
-    taker.task(noop);
+    expect(taker.task('noop')).to.be.undefined();
```
why was this test changed?

sttk replied:
Because I wanted to check that `task` returned nothing when working as a setter.
Is it better to change back to original?

phated commented on test/task.js:
```diff
@@ -24,21 +24,28 @@ describe('task', function() {
 
   it('should register an anonymous function by string name', function(done) {
-    taker.task('test1', anon);
+    expect(taker.task('test1')).to.be.undefined();
```
why was this test changed?

sttk replied:
Because I wanted to check that `task` returned nothing when working as a setter.
Is it better to change back to original?

phated commented on test/task.js:
```diff
@@ -53,7 +60,7 @@ describe('task', function() {
 
   it('should register a named function by string name', function(done) {
-    taker.task('test1', noop);
+    expect(taker.task('test1', noop)).to.be.undefined();
```
why was this test changed?

sttk replied:
Because I wanted to check that `task` returned nothing when working as a setter.
Is it better to change back to original?

phated commented on test/task.js:
```diff
@@ -64,14 +71,18 @@ describe('task', function() {
 
   it('should get a task that was registered', function(done) {
-    taker.task('test1', noop);
+    expect(taker.task('test1', noop)).to.be.undefined();
```
why was this test changed?

sttk replied:
Because I wanted to check that `task` returned nothing when working as a setter.
Is it better to change back to original?

phated commented on test/task.js:
```diff
@@ -64,14 +71,18 @@ describe('task', function() {
 
   it('should return a function that was registered in some other way', function(done) {
-    taker.registry()._tasks.test1 = noop;
+    taker.task('test1', noop);
```
why was this test changed?

sttk replied:
This is mixed the change for the previous modifications.
I'll modify back to original.

phated commented on test/task.js:
```diff
@@ -149,4 +160,67 @@ describe('task', function() {
     done();
   });
 
+  it('use fn.description as a task description', function(done) {
```
I don't think I want undertaker to know about gulp-cli specific properties

sttk replied:
I got it. I'll remove this case.

phated commented on test/task.js:
```diff
@@ -149,4 +160,67 @@ describe('task', function() {
+    done();
+  });
+
+  it('use fn.flag as a task options', function(done) {
```
I don't think I want undertaker to know about gulp-cli specific properties

sttk replied:
I got it. I'll remove this case.

phated commented on test/task.js:
```diff
@@ -149,4 +160,67 @@ describe('task', function() {
+    done();
+  });
+
+  it('take over a description and a flag between tasks', function(done) {
```
I don't think I want undertaker to know about gulp-cli specific properties

sttk replied:
I got it. I'll remove this case.

sttk commented:
Thank you for detail reviews.

I want to move all changes to gulp-cli and make undertaker no change.

Because, though my first plan contains chaining description and control of depth in `tree` function, we removed the former and the latter can be moved to gulp-cli, so it isn't necessary to modify undertaker any more.

I'll implement what you pointed out here into gulp-cli.

About removing `opts.deep` I think it is better to separate.

Can I close this PR about undertaker with no change?

phated commented:
Sounds good!
